### ISSUE TITLE
Add support for synchronous operations

### ISSUE THREAD

**Issue by ibroheem** [Qihoo360/evpp#97]
Take the real world operation below

- Server3 connects to Server1
- Server1 replies Server3
- Server3 use Client (Dispatcher) to communicate with Server2

- Client (Dispatcher) connects and send message to Server2
- Server2 sends reply Client (Dispatcher)
- Client (Dispatcher) gives the message to Server3

- Server3 replies Server1

For Client (Dispatcher) to be possible, I think there should be support for synchronous operations, I have NOT yet able to achieve this using asynchronous

Thanks

![untitled](https://user-images.githubusercontent.com/4117847/32339641-b0a52628-bff8-11e7-9ba0-75511f2c7b93.png)

**Comment by ibroheem** [comment]
The TcpClient class cannot do this current AFAIK

**Comment by zieckey** [comment]
The TcpClient can do this job asynchronously well. I think any synchronous operations can be implemented by asynchronously operations.

**Comment by ibroheem** [comment]
Can u explain please 'cos I tried to, it doesn't

**Comment by ibroheem** [comment]
It might work, but it will be painful to implement. These things just need a send/recv mechanisms.
Everything is NOT async

**Comment by Gabsylar1204** [comment]
Yeah, the same problem. Message come several per one transfer, while supposed to come one at a time. And it seems not to be so friendly with mutexes that much...
